cpu/atmega_common/periph/i2c.c

```diff
@@ -0,0 +1,351 @@
+        /* write data to register */
+        bytes = _write(data, length);
+    }
+    /* finish transfer */
+    _stop();
+    /* return number of bytes send */
+    return bytes;
+}
+
+void i2c_poweron(i2c_t dev)
+{
+    assert(dev < I2C_NUMOF);
+    I2C_POWER_REG &= ~(1 << PRTWI);
+}
+
+void i2c_poweroff(i2c_t dev)
+{
+    assert(dev < I2C_NUMOF);
+    I2C_POWER_REG |= (1 << PRTWI);
+}
+
+static int _start(uint8_t address, uint8_t rw_flag)
+{
+    /* Reset I2C Interrupt Flag and transmit START condition */
+    TWCR = (1 << TWINT) | (1 << TWSTA) | (1 << TWEN);
+    DEBUG("START condition transmitted\n");
+
+    /* Wait for TWINT Flag set. This indicates that the START has been
+    transmitted, and ACK/NACK has been received.*/
+    while (!(TWCR & (1 << TWINT))) {}
+
+    /* Check value of TWI Status Register. Mask prescaler bits.
+    If status different from START go to ERROR */
+    if ((TWSR & 0xF8) == MT_START) {
+        DEBUG("I2C Status is: START\n");
+    }
+    else if ((TWSR & 0xF8) == MT_START_REPEATED) {
+        DEBUG("I2C Status is: START REPEATED\n");
+    }
+    else {
+        DEBUG("I2C Status Register is different from START/START_REPEATED\n");
+        _stop();
+        return -1;
+    }
+
+
+    /* Load ADDRESS and R/W Flag into TWDR Register.
+    Clear TWINT bit in TWCR to start transmission of ADDRESS */
+    TWDR = (address << 1) | rw_flag;
+    TWCR = (1 << TWINT) | (1 << TWEN);
+    DEBUG("ADDRESS and FLAG transmitted\n");
+
+    /* Wait for TWINT Flag set. This indicates that ADDRESS has been transmitted.*/
+    while (!(TWCR & (1 << TWINT))) {}
+
+    /* Check value of TWI Status Register. Mask prescaler bits.
+    If status different from ADDRESS ACK go to ERROR */
+    if ((TWSR & 0xF8) == MT_ADDRESS_ACK) {
+        DEBUG("ACK has been received for ADDRESS (write)\n");
+    }
+    else if ((TWSR & 0xF8) == MR_ADDRESS_ACK) {
+        DEBUG("ACK has been received for ADDRESS (read)\n");
+    }
+    else {
+        DEBUG("NOT ACK has been received for ADDRESS\n");
+        _stop();
+        return -2;
+    }
+
+    return 0;
+}
+
+static int _write(const uint8_t *data, int length)
+{
+    int i = 0;
+
+    for (i = 0; i < length; i++) {
+        /* Load DATA into TWDR Register.
+        Clear TWINT bit in TWCR to start transmission of data */
+        TWDR = data[i];
+        TWCR = (1 << TWINT) | (1 << TWEN);
+        DEBUG("Byte %i transmitted\n", i+1);
+
+        /* Wait for TWINT Flag set. This indicates that DATA has been transmitted.*/
+        while (!(TWCR & (1 << TWINT))) {}
+
+        /* Check value of TWI Status Register. Mask prescaler bits. If status
+        different from MT_DATA_ACK, return number of transmitted bytes */
+        if ((TWSR & 0xF8) != MT_DATA_ACK) {
+            DEBUG("NACK has been received for BYTE %i\n", i+1);
+            return i;
+        }
+        else {
+            DEBUG("ACK has been received for BYTE %i\n", i+1);
+        }
+    }
+
+    return i;
```

Comment:
if you get here, its save to return length, or not? So `i` is not needed here, as suggested above